packages/components/src/internal/components/files/FileAttachmentContainer.tsx

```diff
@@ -92,10 +98,59 @@ export class FileAttachmentContainer extends React.Component<
         });
     };
 
+    getFilesFromDirectory = (files, entry, scope, callback): void => {
```

Comment:
This method is lacking in typings. Was this copied from somewhere else? Some left-hand typings do help a bit but also it is helpful here to use the "is" pattern to recognize types at run time. Here are a couple of utilities:

```tsx
const isDirectoryEntry = (entry: FileSystemEntry): entry is FileSystemDirectoryEntry => entry.isDirectory;
const isFileEntry = (entry: FileSystemEntry): entry is FileSystemFileEntry => entry.isFile;
```

I was able to get it pretty much fully typed:

```tsx
getFilesFromDirectory = (
    files: Record<string, File>,
    entry: FileSystemDirectoryEntry,
    scope: FileAttachmentContainer,
    callback: () => void
): void => {
    const dirReader: FileSystemDirectoryReader = entry.createReader();

    const entriesReader: FileSystemEntriesCallback = (scope => {
        return entries => {
            for (let i = 0; i < entries.length; i++) {
                const _entry = entries[i];
                if (isFileEntry(_entry)) {
                    scope.fileCbCount++;
                    _entry.file(file => {
                        scope.fileCbCount--;

                        // ignore hidden files
                        if (file.name.substring(0, 1) !== '.') {
                            if (files[file.name]) {
                                scope.setState({
                                    warningMsg:
                                        'Duplicate files were uploaded. Only the last file provided for the duplicate name will be included in the file set.',
                                });
                            }
                            // @ts-expect-error fullPath is not a property of File in lib.dom.ts
                            file.fullPath = entry.fullPath + '/' + file.name;
                            files[file.name] = file;
                        }

                        if (scope.dirCbCount === 0 && scope.fileCbCount === 0) {
                            callback();
                        }
                    });
                } else if (isDirectoryEntry(_entry)) {
                    scope.getFilesFromDirectory(files, _entry, scope, callback);
                }
            }

            // readEntries only reads 100 files in a batch for Chrome
            if (entries.length >= 100) {
                scope.dirCbCount++;
                dirReader.readEntries(entriesReader, console.error);
            }

            scope.dirCbCount--;
            if (scope.dirCbCount === 0 && scope.fileCbCount === 0) {
                callback();
            }
        };
    })(scope);

    scope.dirCbCount++;
    dirReader.readEntries(entriesReader, console.error);
};
```

Reply:
Awesome. Thanks. This was copied (and extended a bit) from the LABKEY.internal.ZipLoad.getFilesFromDirectory function in ZipLoad.js.